lib/elixir/unicode/unicode.ex

```diff
@@ -305,70 +331,135 @@ defmodule String.Casing do
                                   to_binary.(title)})
   end
 
-  # Downcase
+  for {codepoint, upper, lower, title} <- codes do
+    def get_codepoint_info(unquote(codepoint)) do
+      if is_nil(unquote(lower)) || unquote(codepoint) == unquote(lower) do
+        %{case: :lower, upper: unquote(upper),
+                        lower: unquote(codepoint),
+                        title: unquote(title)}
+      else
+        %{case: :upper, upper: unquote(codepoint),
+                        lower: unquote(lower),
+                        title: unquote(title)}
+      end
+    end
+  end
+  
+  def get_codepoint_info(_), do: %{}
 
-  def downcase(string), do: downcase(string, "")
+  # Downcase
 
+  def downcase_orig(string), do: downcase_orig(string, "")
+  
   for {codepoint, _upper, lower, _title} <- codes, lower && lower != codepoint do
-    defp downcase(unquote(codepoint) <> rest, acc) do
-      downcase(rest, acc <> unquote(lower))
+    defp downcase_orig(unquote(codepoint) <> rest, acc) do
+      downcase_orig(rest, acc <> unquote(lower))
     end
   end
-
-  defp downcase(<<char, rest::binary>>, acc) do
-    downcase(rest, <<acc::binary, char>>)
+  
+  defp downcase_orig(<<char, rest::binary>>, acc) do
+    downcase_orig(rest, <<acc::binary, char>>)
   end
+  
+  defp downcase_orig("", acc), do: acc
 
+
+  def downcase(string), do: downcase(string, "")
+    
+  defp downcase(<<char :: utf8, rest::binary>>, acc) do
+    ch_down = case get_codepoint_info(<<char :: utf8>>) do
```

Comment:
Cool. Can you please try a foreign text as well? Maybe something in
Japanese or Korean?

On Wednesday, June 29, 2016, Vadim Agishev <notifications@github.com> wrote:

> In lib/elixir/unicode/unicode.ex
> <https://github.com/elixir-lang/elixir/pull/4899#discussion_r68930383>:
>
> >
> > +
> > +  def downcase(string), do: downcase(string, "")
> > +
> > +  defp downcase(<<char :: utf8, rest::binary>>, acc) do
> > +    ch_down = case get_codepoint_info(<<char :: utf8>>) do
>
> Yea, faster.
> downcase was 2,47x slower, now 2,08x
> upcase was 2,30x slower, now 1,6x
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/4899/files/abeb3bd05f6abe4be1352291dfb90164e6093f70#r68930383>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAlbrn1pIy6Y_Ic-LpFaheUXpWumXFqks5qQlexgaJpZM4I_T9c>
> .
>


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*